Support non-quadratic viewports

philippotto opened the pull request:
### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- test protocol will be done by the lab

### Issues:
- fixes #3522 

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [X] Ready for review

philippotto commented on app/assets/javascripts/oxalis/model/accessors/flycam_accessor.js:
```diff
@@ -199,12 +203,13 @@ export function getRequestLogZoomStep(state: OxalisState): number {
   return clamp(min, qualityAdaptedZoomStep, maxLogZoomStep);
 }
 
-export function getTextureScalingFactor(state: OxalisState): number {
```
This function wasn't used anymore.

philippotto commented:
@daniel-wer I think it makes sense if you start reviewing this PR. There are some parts which are not spotless (bucket calculation for arbitrary mode and there are some code fragments for an rectangular arbitrary view which are not effective). However, the main part should be ready for review.

Also: Maybe you have an idea on how to estimate the bucket count which is needed for arbitrary view. Right now, I switched back to the very old way of calculating the zoomstep (`log(zoomValue / 1.3)`), which is not exact.

daniel-wer commented on app/assets/javascripts/oxalis/controller/scene_controller.js:
```diff
@@ -270,7 +272,7 @@ class SceneController {
   }
 
-  updateSceneForCam = (id: OrthoView): void => {
+  updateSceneForCam = (id: OrthoView, hidePlanes: boolean = false): void => {
```
Is this argument supplied somewhere? Couldn't find it.

daniel-wer commented on app/assets/javascripts/oxalis/geometries/arbitrary_plane.js:
```diff
@@ -100,6 +112,13 @@ class ArbitraryPlane {
 
+        mesh.matrix.identity();
+
+        // const [xFactor, yFactor] = applyAspectRatioToWidth(this.aspectRatio, 1);
```
Not sure whether you want to use the commented out code in the future, otherwise this can probably be reverted to the way it was before.

daniel-wer commented on app/assets/javascripts/oxalis/geometries/materials/node_shader.js:
```diff
@@ -37,7 +37,11 @@ class NodeShader {
       planeZoomFactor: {
         type: "f",
-        value: getPlaneScalingFactor(state.flycam),
+        // The flycam zoom is typically decomposed into an x- and y-factor
```
:+1:

daniel-wer commented on app/assets/javascripts/oxalis/constants.js:
```diff
@@ -111,10 +112,16 @@ export const POSITION_REF_REGEX = /#\(([0-9]+,[0-9]+,[0-9]+)\)/g;
+// Using the following dimensions for the address space,
+// the look up buffer (256**2) is used at a rate of ~ 97%
+// ((32 × 32 × 50 + 16 × 16 × 50) / 256^2 = 0.976563)
+export const addressSpaceDimensions = {
```
Just to understand correctly: This means from the anchor point on, at most 32 buckets in x/y direction and 50 buckets in z direction can be put into the lookup buffer?

philippotto replied:
Yes, this is correct. If/when we convert this code to hashing at some point, these limitation would be lifted.

daniel-wer commented on app/assets/javascripts/oxalis/model/accessors/flycam_accessor.js:
```diff
@@ -52,14 +50,25 @@ function _approximateMaxZoomForZoomStep(
+  // Since the viewports can be quite large, it can happen that even a zoom value of 1 is not feasible.
+  // That's why we start the search with a smaller value than 1. We use the ZOOM_STEP_INTERVAL factor
+  // to ensure that the calculated thresholds correspond to the normal zoom behavior.
+  let maxZoomStep = 1 / ZOOM_STEP_INTERVAL ** 20;
```
Because you're starting 20 iterations earlier now, we only search up until 80 iterations instead of the 100 that were the maximum before (if that's understandable :see_no_evil:) Maybe we should increase `maximumIterationCount` by 20 then?

philippotto replied:
Good point! I guess, I will rewrite this in a new branch to use a proper end criteria (which is dependent on the resolution count) as we discussed.

daniel-wer commented on app/assets/javascripts/oxalis/model/accessors/view_mode_accessor.js:
```diff
@@ -1,17 +1,19 @@
 } from "oxalis/constants";
 
-export function getTDViewportSize(): number {
-  // the viewport is always quadratic
```
:grinning:

daniel-wer commented on app/assets/javascripts/oxalis/store.js:
```diff
@@ -421,7 +421,7 @@ export const defaultState: OxalisState = {
     crosshairSize: 0.1,
     displayCrosshair: true,
-    displayScalebars: false,
+    displayScalebars: true,
```
:+1:

daniel-wer commented on app/assets/javascripts/oxalis/view/input_catcher.js:
```diff
@@ -21,20 +21,29 @@ function ignoreContextMenu(event: SyntheticInputEvent<>) {
 // makes the input catcher a square and returns its position within the document
 // relative to the rendering canvas
-function makeInputCatcherQuadratic(inputCatcherDOM: HTMLElement): Rect {
+function makeInputCatcherQuadratic(inputCatcherDOM: HTMLElement, makeQuadratic: boolean): Rect {
```
The function name and comment should be adjusted to more closely reflect what the function is doing now

daniel-wer commented on app/assets/javascripts/oxalis/view/layouting/default_layout_configs.js:
```diff
@@ -17,7 +17,7 @@ export const currentLayoutVersion = 7;
 export const headerHeight = 55;
 const dummyExtent = 500;
-export const show3DViewportInArbitrary = false;
+export const show3DViewportInArbitrary = true;
```
Deliberate change?

daniel-wer commented on app/assets/javascripts/oxalis/view/right-menu/dataset_info_tab_view.js:
```diff
@@ -84,12 +84,11 @@ const shortcuts = [
-  if (constants.MODES_PLANE.includes(viewMode)) {
-    width = constants.PLANE_WIDTH;
-  } else if (constants.MODES_ARBITRARY.includes(viewMode)) {
+  if (constants.MODES_PLANE.includes(viewMode) || constants.MODES_ARBITRARY.includes(viewMode)) {
```
We can probably get rid of this whole if/else and simply replace it with `width = constants.VIEWPORT_WIDTH;`

daniel-wer commented on app/assets/javascripts/oxalis/view/right-menu/dataset_info_tab_view.js:
```diff
@@ -84,12 +84,11 @@ const shortcuts = [
 
 export function calculateZoomLevel(flycam: Flycam, dataset: APIDataset): number {
-  const zoom = getPlaneScalingFactor(flycam);
+  // todo
```
What is this todo about?

philippotto replied:
Hmm, good point, I forgot about that. The viewport extent in nm is not shown right now (since it differs per viewport edge). Will have to find a solution for that, yet.

philippotto commented on app/assets/javascripts/oxalis/model/bucket_data_handling/bucket_picker_strategies/orthogonal_bucket_picker.js:
```diff
@@ -42,31 +42,54 @@ import { extraBucketPerEdge, extraBucketsPerDim } from "./orthogonal_bucket_pick
   return addExtraBuckets(bucketCountPerDim);
 }
 
+function calculateBucketCountPerViewportEdge(
```
Yeah, the naming is weird, but luckily this will soon be gone hopefully :)

philippotto commented on app/assets/javascripts/oxalis/model/bucket_data_handling/texture_bucket_manager.js:
```diff
@@ -300,21 +292,25 @@ export default class TextureBucketManager {
-    if (x < 0) console.warn("x should be greater than 0. is currently:", x);
-    if (y < 0) console.warn("y should be greater than 0. is currently:", y);
-    if (z < 0) console.warn("z should be greater than 0. is currently:", z);
+    // if (x < 0) console.warn("x should be greater than 0. is currently:", x);
```
No, not really, but it's a bit of a performance killer (the actual log statements). I plan to re-investigate it soon.

philippotto commented:
I fixed the scalebars and added a tooltip for the whole viewport extent (since this isn't in the dataset info tab view anymore).

![image](https://user-images.githubusercontent.com/2486553/52484103-d1c60000-2bb5-11e9-963c-07f32417abb2.png)

daniel-wer commented on app/assets/javascripts/oxalis/model/bucket_data_handling/bucket_picker_strategies/orthogonal_bucket_picker.js:
```diff
@@ -283,6 +116,8 @@ function addNecessaryBucketsToPriorityQueueOrthogonal(
     const additionalPriorityWeight = isFallback ? 1000 : 0;
 
     // Build up priority queue
+    // todo
```
Do you want to do this as part of this PR?

philippotto replied:
I think this "todo" is obsolete. At least, I cannot see anything missing here. Do you?

daniel-wer commented on app/assets/javascripts/oxalis/model/bucket_data_handling/bucket_picker_strategies/flight_bucket_picker.js:
```diff
@@ -149,12 +147,13 @@ export default function determineBucketsForFlight(
   const fallbackBuckets = isFallbackAvailable ? traverseFallbackBBox(getBBox(planeBuckets)) : [];
   traversedBuckets = traversedBuckets.concat(fallbackBuckets);
 
+  let currentCount = 0;
```
The `const uniqueBucketMap = new ThreeDMap();` is used for the other two bucket pickers but not this one, is there a reason for that?

philippotto replied:
"Historical growth" :see_no_evil: I forgot about the approach in the flight bucket picker (which was there before) and used the ThreeDMap instead. I'll probably consolidate this when doing the arbitrary-fallback-rendering PR.

daniel-wer commented on app/assets/javascripts/oxalis/model/bucket_data_handling/bucket_picker_strategies/orthogonal_bucket_picker.js:
```diff
@@ -296,24 +131,27 @@ function addNecessaryBucketsToPriorityQueueOrthogonal(
             x === extraXBucketStart ||
             x === extraXBucketEnd;
 
-          if (bucket.type !== "null") {
```
So the slight disadvantage of the new method is that buckets outside of the bounding box count against the "quota", because we no longer know in this spot whether they are null buckets or not?

philippotto replied:
Absolutely right! However, the previous approach couldn't take into account "out of bbox"-buckets, either, since it calculated the worst-case amount. Also, the average use case shouldn't have any "out of bbox"-buckets, so, the approach should be fine.

daniel-wer commented on flow-typed/npm/js-priority-queue_vx.x.x.js:
```diff
@@ -0,0 +1,29 @@
+ * https://github.com/flowtype/flow-typed
+ */
+
+declare module "js-priority-queue" {
```
Awesome!

philippotto replied:
:innocent:

philippotto commented on .circleci/not-on-master.sh:
```diff
@@ -1,8 +1,8 @@
 set -Eeuo pipefail
 
-if [ "${CIRCLE_BRANCH}" == "master" ]; then
+# if [ "${CIRCLE_BRANCH}" == "master" ]; then
```
TODO: This needs to be undone before merging.
